Add plexon2 support

JuliaSprenger opened the pull request:
This is a draft to add support for the new Plexon pl2 support to neo as among other sources requested in https://github.com/NeuralEnsemble/python-neo/issues/1124

The RawIO will be based on the PyPl2 scripts and corresponding dlls provided by Plexon (see https://gitlab.com/penalab/pypl2 and https://github.com/juliasprenger/pypl2)

This PR requires the test file to be available on GIN, so https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/pulls/97/files needs to be merged first.

JuliaSprenger commented:
This PR is in principle ready for review. Test files need to be merged for tests to pass.

I didn't have any test files with multiple analog signal fragments present, if anybody has a file with this feature it would be great to be tested before merging.

@brian-lau @KonKob @vincisLab

JuliaSprenger commented:
TODOs:
- [ ] Move the dll to an external repo and fetch it.

bendichter commented on neo/rawio/plexon2rawio/plexon2rawio.py:
why have the default value of filename be an empty string? Shouldn't this be a required arg?

bendichter commented on neo/rawio/plexon2rawio/plexon2rawio.py:
Maybe this should live in the package directory. That way when you uninstall the package you uninstall the dll

JuliaSprenger replied:
This approach might cause problems if the installation was performed with different right than the execution. The package directory might be read-only. Creating a folder for the dll in the home directory should not fail due to permissions and is the same location used also for cloning of the `ephy_testing_files` repository. I agree that there should be a mechanism to register these additional folders in the pyproject.toml to be also removed upon deinstallation of neo, but didn't find the right option yet. Let me know if you have seen something like this already.

JuliaSprenger commented:
Tests are not completing at the moment.
- [ ] fix Plexon2IO tests

JuliaSprenger commented:
@samuelgarcia The failing tests are not related to the new IO. Could you have a look at the proposed IO?

JuliaSprenger commented:
Hi @samuelgarcia Ready for review!

samuelgarcia commented on neo/rawio/plexon2rawio/plexon2rawio.py:
I think we can also use `https://raw.githubusercontent.com/...` no ?

JuliaSprenger replied:
Yes, you are right. This way we can remove the `/blob` and `?raw=true`
```suggestion
                dist = urlopen('https://raw.githubusercontent.com/Neuralensemble/pypl2/master/bin/PL2FileReader.dll')
```

samuelgarcia commented on neo/rawio/plexon2rawio/plexon2rawio.py:
This comments is from the example no ?

JuliaSprenger replied:
Yes, i kept some of the comments from the example io, but this one is not applicable any more. I removed it.

samuelgarcia commented on neo/rawio/plexon2rawio/plexon2rawio.py:
I guess `pl2_get_analog_channel_data_by_name` is very slow...

JuliaSprenger replied:
I did some test runs of this on a multi-GB pl2 file and it seems that the `_by_name` method is even a tiny bit faster than accessing the channel by its id. I don't know what exactly is happening in the back as the dll is closed source, but I would stick to the `pl2_get_analog_channel_dat_by_name` method here for now.

samuelgarcia commented on neo/rawio/plexon2rawio/plexon2rawio.py:
To be safe here if we correct a bug one day I would share this piece of code with the one in _get_spike_timestamps

samuelgarcia commented on neo/rawio/plexon2rawio/plexon2rawio.py:
I think I would do this on demand and not in init.

JuliaSprenger replied:
Good idea, I moved the loading into cache in the respective methods.

samuelgarcia commented:
Hi julia.
Thank you for this incredibly tedious task of wrapping a windows ddl here.
This is heroic.
My guess is that implementing from scratch the reader with good specification would be simpler :)

This is really hard to really give feedback of the ctype wrapper layer because I do not known the ddl.
I made some generic feedback a some places you can fix or not.

Maybe you could do some clean on the comments hinerited from exmaplerawio to keep only relevent ones.

JuliaSprenger commented on neo/rawio/plexon2rawio/plexon2rawio.py:
Done.

JuliaSprenger commented on neo/rawio/plexon2rawio/plexon2rawio.py:
Good catch, I fixed it by adding a conversion factor.

samuelgarcia commented:
Hi Julia.
Tell me when ready for merge.

JuliaSprenger commented:
Hi @samuelgarcia Thanks for the feedback. I addressed your comments and in principle only one is still open regarding loading of data by the channel name. I cleaned some of the comments that remained from the example io, but still many of them might be useful later on to understand why certain methods have certain return values and signatures. Can you check again and see if you are happy with the current changes? From my side this is ready for merging.

samuelgarcia commented:
OK lets go go for merging.
@bendichter @Chetank99  and @JuliaSprenger : total apologies for having so much delays.

Chetank99 commented on neo/rawio/plexon2rawio/plexon2rawio.py:
Hi @JuliaSprenger 
This pull request was super helpful!!
Just one thing, while executing Plexon2RawIO(). I got the error:
OSError: [WinError 193] %1 is not a valid Win32 application
and
OSError: Error: Can't load PL2FileReader.dll at: C:\Users\admin.plexon_dlls_for_neo\PL2FileReader.dll. PL2FileReader.dll is bundled with the C++ PL2 Offline Files SDK located on the Plexon Inc website: www.plexon.com Contact Plexon Support for more information: support@plexon.com

I then realized my pypl2 SDK folder had another dll file called "PL2FileReader64.dll"
so after replace here "PL2FileReader.dll" with "PL2FileReader64.dll", it started running smoothly.

Just wanted to keep a note here if anyone else encounters those errors and wants to move ahead
